server/src/test/java/org/elasticsearch/cluster/NodeConnectionsServiceTests.java

```diff
@@ -77,65 +87,212 @@
         return nodes;
     }
 
-    private ClusterState clusterStateFromNodes(List<DiscoveryNode> nodes) {
+    private DiscoveryNodes discoveryNodesFromList(List<DiscoveryNode> discoveryNodes) {
         final DiscoveryNodes.Builder builder = DiscoveryNodes.builder();
-        for (DiscoveryNode node : nodes) {
-            builder.add(node);
+        for (final DiscoveryNode discoveryNode : discoveryNodes) {
+            builder.add(discoveryNode);
         }
-        return ClusterState.builder(new ClusterName("test")).nodes(builder).build();
+        return builder.build();
     }
 
-    public void testConnectAndDisconnect() {
-        List<DiscoveryNode> nodes = generateNodes();
-        NodeConnectionsService service = new NodeConnectionsService(Settings.EMPTY, threadPool, transportService);
-
-        ClusterState current = clusterStateFromNodes(Collections.emptyList());
-        ClusterChangedEvent event = new ClusterChangedEvent("test", clusterStateFromNodes(randomSubsetOf(nodes)), current);
-
-        service.connectToNodes(event.state().nodes());
-        assertConnected(event.state().nodes());
+    public void testConnectAndDisconnect() throws Exception {
+        final NodeConnectionsService service = new NodeConnectionsService(Settings.EMPTY, threadPool, transportService);
 
-        service.disconnectFromNodesExcept(event.state().nodes());
-        assertConnectedExactlyToNodes(event.state());
+        final AtomicBoolean stopReconnecting = new AtomicBoolean();
+        final Thread reconnectionThread = new Thread(() -> {
+            while (stopReconnecting.get() == false) {
+                final PlainActionFuture<Void> future = new PlainActionFuture<>();
+                service.ensureConnections(() -> future.onResponse(null));
+                future.actionGet();
+            }
+        });
+
+        final List<DiscoveryNode> allNodes = generateNodes();
+        for (int iteration = 0; iteration < 3; iteration++) {
+            final DiscoveryNodes nodes = discoveryNodesFromList(randomSubsetOf(allNodes));
+
+            final PlainActionFuture<Void> future = new PlainActionFuture<>();
+            service.connectToNodes(nodes, () -> future.onResponse(null));
+            future.actionGet();
+            assertConnected(nodes);
+            service.disconnectFromNodesExcept(nodes);
+
+            if (randomBoolean()) {
+                // sometimes do not wait for the disconnections to complete before starting the next connections
```

Comment:
Given disconnect is blocking, this will not really start connecting before disconnection is complete (thought above thread might)?

Reply:
A hangover from a time when disconnections were also done in the background. I added some background disconnections in 371c263d8b2d6d2be7fe635e5c0008d7271a2095 to give this assertion some teeth again: the `disruptionThread` might disrupt the connection to a node, then the `reconnectionThread` attempts to re-establish it, then concurrently we try and remove the node using `service.disconnectFromNodesExcept()`, and in this case we do not wait for the disconnection to complete.